gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,972 @@
+
+  if (!g_spawn_async_with_pipes (NULL, args, NULL,
+                                 G_SPAWN_SEARCH_PATH | G_SPAWN_DO_NOT_REAP_CHILD,
+                                 NULL, NULL, &reformatter->gpg,
+                                 &reformatter->gpg_in, NULL, NULL, &error))
+    {
+      reformatter->error = g_error_new (EOS_REFORMATTER_ERROR, 0, _("Image verification error."));
+      return FALSE;
+    }
+  else
+    {
+      g_child_watch_add (reformatter->gpg, (GChildWatchFunc)eos_reformatter_gpg_watch, reformatter);
+    }
+
+  return TRUE;
+}
+
+/* Thread runners */
+
+static gpointer do_reformat (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean read_ok = TRUE;
+  gboolean decomp_ok = TRUE;
+  gboolean write_ok = TRUE;
+
+  do
+    {
+      if (read_ok)
+        read_ok = eos_reformatter_read (reformatter, reformatter->write_queue);
+      if (reformatter->decompressor != NULL && decomp_ok)
+        decomp_ok = eos_reformatter_decompress (reformatter, reformatter->decomp_queue, reformatter->write_queue);
+      write_ok = eos_reformatter_write (reformatter, reformatter->write_queue);
+    }
+  while (read_ok || write_ok || decomp_ok);
+
+  return NULL;
+}
+
+static gpointer do_read (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean ok = TRUE;
+
+  do
+    {
+      if (reformatter->decomp_thread == NULL)
+        {
+          ok = eos_reformatter_read (reformatter, reformatter->write_queue);
+        }
+      else
+        {
+          ok = eos_reformatter_read (reformatter, reformatter->decomp_queue);
+        }
+    }
+  while (ok);
+
+  return NULL;
+}
+
+static gpointer do_decompress (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean ok = TRUE;
+
+  do
+    {
+      ok = eos_reformatter_decompress (reformatter, reformatter->decomp_queue, reformatter->write_queue);
+    }
+  while (ok);
+
+  return NULL;
+}
+
+static gpointer do_read_and_decompress (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean read_ok = TRUE;
+  gboolean decomp_ok = TRUE;
+
+  do
+    {
+      if (read_ok)
+        read_ok = eos_reformatter_read (reformatter, reformatter->decomp_queue);
+
+      decomp_ok = eos_reformatter_decompress (reformatter, reformatter->decomp_queue, reformatter->write_queue);
+    } while (read_ok || decomp_ok);
+
+  return NULL;
+}
+
+static gpointer do_write (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean ok = TRUE;
+
+  do
+    {
+      ok = eos_reformatter_write (reformatter, reformatter->write_queue);
+    } while (ok);
+
+  return NULL;
+}
+
+
+/* Public API */
+
+gdouble
+eos_reformatter_get_progress (EosReformatter *reformatter)
+{
+  gdouble progress = 0.0;
+  g_object_get (reformatter, "progress", &progress, NULL);
+  return progress;
+}
+
+GError *
+eos_reformatter_get_error (EosReformatter *reformatter)
+{
+  return reformatter->error;
```

Comment:
Yeah, the API should be changed to deliver the error through something better than a simple getter.